src/pwrite.rs

```diff
@@ -24,12 +24,11 @@ use crate::error;
 /// have a data that is not `&[u8]` — you will probably want to implement
 /// [TryIntoCtx](ctx/trait.TryIntoCtx.html) on your Rust types to be written.
 ///
-pub trait Pwrite<Ctx, E>: Index<usize> + IndexMut<RangeFrom<usize>> + MeasureWith<Ctx>
-where
-    Ctx: Copy,
-    E: From<error::Error>,
-{
-    fn pwrite<N: TryIntoCtx<Ctx, <Self as Index<RangeFrom<usize>>>::Output, Error = E>>(
+pub trait Pwrite<Ctx: Copy, E> {
+    /// The type used by implementors of `TryIntoCtx` to write into your data store.
+    type Buffer: ?Sized;
```

Comment:
so this part is peculiar to me; i have to think about it more; specifically, I'm wondering why this associated type is required; in e.g., your impl Buffer is Self, and pwrite requires basically `TryIntoCtx<Ctx, Self::Buffer, Error>`, which means `TryIntoCtx<Ctx, Self, Error>` so why wouldn't the associated type be dropped and we just pass Self there always?

I'm sure there's a good reason, but it would be nice if possible not to introduce the associated type, though if it is requried I think it's fine. I don't know of anyone putting explicit `Pwrite` bounds on their generic types (it shoudl only require `TryIntoCtx` bounds in order for `Pwrite` to work.

Quite an interesting solution to the problem you've done!

Reply:
That's a good point, and I think you might be right. I added an associated type to make sure that the blanket implementations that already existed wouldn't change behavior, but if we change that to just a `impl [u8]` or something, then that shouldn't be a problem and we can get rid of it.